File: app/src/main/java/com/bing/example/main/fragment/VideoListFragment.java
package com.bing.example.main.fragment;

import androidx.annotation.NonNull;
import androidx.core.content.FileProvider;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import cn.jzvd.JzvdStd;

import android.content.Intent;
import android.net.Uri;
import android.view.View;

import com.bing.example.R;
import com.bing.example.databinding.FragmentVideoListBinding;
import com.bing.example.main.adapter.VideosAdapter;
import com.bing.example.main.viewmodel.MainViewModel;
import com.bing.example.main.viewmodel.VideoListViewModel;
import com.bing.example.model.RepositoryManager;
import com.bing.example.model.entity.VideoInfo;
import com.bing.mvvmbase.base.recycleview.BaseRecycleViewFragment;
import com.bing.mvvmbase.model.datawrapper.Status;
import com.blankj.utilcode.util.FileUtils;
import com.blankj.utilcode.util.SizeUtils;
import com.blankj.utilcode.util.ToastUtils;
import com.scwang.smartrefresh.layout.SmartRefreshLayout;
import com.scwang.smartrefresh.layout.api.RefreshLayout;
import com.yarolegovich.lovelydialog.LovelyStandardDialog;
import com.yarolegovich.lovelydialog.LovelyTextInputDialog;

import java.io.File;
import java.util.List;

import static com.bing.example.main.adapter.VideosAdapter.MODE_NORMAL;
import static com.bing.example.main.adapter.VideosAdapter.MODE_SELECT;

public class VideoListFragment extends BaseRecycleViewFragment<FragmentVideoListBinding, VideoListViewModel, MainViewModel, VideosAdapter, VideoInfo>  {

        public VideoListFragment() {
                // Required empty public constructor
        }

        @Override
        protected void handleArguments() {

        }

        @Override
        protected int layoutId() {
                return R.layout.fragment_video_list;
        }

        @Override
        protected void initActivityViewModel() {
                mActivityViewModel = ViewModelProviders.of(getActivity()).get(MainViewModel.class);
        }

        @Override
        protected void initViewModel() {
                mViewModel = ViewModelProviders.of(this).get(VideoListViewModel.class);
        }

        @Override
        protected void reload(View view) {

        }

        @Override
        protected SmartRefreshLayout getRefreshLayout() {
                return mBinding.refreshLayout;
        }

        @Override
        protected void initRefreshLayout() {
                super.initRefreshLayout();
                getRefreshLayout().setEnableRefresh(false);
        }

        @Override
        protected void refresh(@NonNull RefreshLayout refreshLayout) {

        }

        @Override
        protected RecyclerView getRecyclerView() {
                return mBinding.recyclerView;
        }

        @Override
        protected RecyclerView.LayoutManager getLayoutManager() {
                return new GridLayoutManager(getContext(), 2);
        }

        @Override
        protected void initRecycleView() {
                super.initRecycleView();
                GridOffsetsItemDecoration offsetsItemDecoration = new GridOffsetsItemDecoration(
                                GridOffsetsItemDecoration.GRID_OFFSETS_VERTICAL);
                offsetsItemDecoration.setVerticalItemOffsets(SizeUtils.dp2px(8));
                offsetsItemDecoration.setHorizontalItemOffsets(SizeUtils.dp2px(8));
                mRecyclerView.addItemDecoration(offsetsItemDecoration);
        }

        @Override
        protected void initAdapter() {
                mAdapter = new VideosAdapter(new VideosAdapter.OnClickListener() {
                        @Override
                        public void onClickImage(int position, VideoInfo videoInfo) {
                                String fileName = videoInfo.getPath();
                                if (FileUtils.isFileExists(fileName)) {
                                        JzvdStd.startFullscreen(getContext(), JzvdStd.class, videoInfo.getPath(), videoInfo.getTitle());
                                } else {
                                        ToastUtils.showShort(R.string.file_not_exits);
                                }
                        }

                        @Override
                        public void onClickRename(int position, VideoInfo videoInfo) {
                                new LovelyTextInputDialog(getContext())
                                                .setTopColorRes(R.color.color18)
                                                .setTitle(R.string.rename)
                                                .setMessage(R.string.please_input_filename)
                                                .setIcon(R.drawable.ic_edit_black_24dp)
                                                .setConfirmButton(android.R.string.ok, text -> {
                                                        videoInfo.setTitle(text);
                                                        RepositoryManager.instance().updateVideo(videoInfo);
                                                })
                                                .show();
                        }

                        @Override
                        public void onClickDelete(int position, VideoInfo videoInfo) {
                                new LovelyStandardDialog(getContext(), LovelyStandardDialog.ButtonLayout.HORIZONTAL)
                                                .setTopColorRes(R.color.color21)
                                                .setButtonsColorRes(R.color.color22)
                                                .setIcon(R.drawable.ic_delete_black_24dp)
                                                .setTitle(R.string.delete)
                                                .setMessage(R.string.sure_delete)
                                                .setPositiveButton(android.R.string.ok, v ->{
                                                        mViewModel.deleteVideo(videoInfo);
                                                })
                                                .setNegativeButton(android.R.string.no, null)
                                                .show();
                        }

                        @Override
                        public void onClickShare(int position, VideoInfo videoInfo) {
                                Uri uri = FileProvider.getUriForFile(getContext(), "com.bing.example.fileprovider", new File(videoInfo.getPath()));

                                Intent shareIntent = new Intent();
                                shareIntent.setAction(Intent.ACTION_SEND);            //分享视频只能单个分享
                                shareIntent.putExtra(Intent.EXTRA_STREAM, uri);
                                shareIntent.setType("audio/*");
                                startActivity(Intent.createChooser(shareIntent, getString(R.string.share_to)));
                        }

                        @Override
                        public void onClickFore(int position, int selectedNumber) {

                        }
                }, mode -> {
                        if (mode == MODE_NORMAL) {
                                mActivityViewModel.getIsNormalMode().set(true);
                        } else {
                                mActivityViewModel.getIsNormalMode().set(false);
                        }
                });
        }

        @Override
        protected MutableLiveData<Status> getNetworkState() {
                return mViewModel.getLoadStatus();
        }

        @Override
        protected MutableLiveData<Status> getRefreshState() {
                return mViewModel.getLoadStatus();
        }

        @Override
        protected LiveData<List<VideoInfo>> getData() {
                return mViewModel.getVideoInfos();
        }

        @Override
        protected void bindAndObserve() {
                super.bindAndObserve();
        }

        public void onClickBack() {
                mAdapter.intoNormalMode();
                mActivityViewModel.getIsNormalMode().set(true);
        }

        public void onClickDelete() {
                mViewModel.deleteVideos(mAdapter.getSelectedData());
                mAdapter.intoNormalMode();
                mActivityViewModel.getIsNormalMode().set(true);
        }

        public void onClickSelectAll() {
                if (mAdapter.getSelectedPostions().size() == mAdapter.getItemCount()) {
                        mAdapter.getSelectedPostions().clear();
                } else {
                        for (int i = 0; i < mAdapter.getData().size(); i++) {
                                mAdapter.getSelectedPostions().add(i);
                        }
                }
                mAdapter.notifyDataSetChanged();
        }

        public boolean onBackPressed() {
                if (mAdapter.getCurrentMode() == MODE_SELECT) {
                        mAdapter.intoNormalMode();
                        return true;
                } else {
                        return false;
                }
        }

        public void intoNormalMode() {
                if (mAdapter.getCurrentMode() == MODE_SELECT) {
                        mAdapter.intoNormalMode();
                }
        }
}

File: app/src/main/java/com/bing/example/main/videolist/VideoListFragment.kt
package com.bing.example.main.videolist

import androidx.core.content.FileProvider
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import app.dinus.com.itemdecoration.GridOffsetsItemDecoration
import cn.jzvd.JzvdStd

import android.content.Intent
import androidx.lifecycle.ViewModelProviders
import com.afollestad.materialdialogs.MaterialDialog
import com.afollestad.materialdialogs.input.input

import com.bing.example.R
import com.bing.example.databinding.FragmentVideoListBinding
import com.bing.example.main.home.MainViewModel
import com.bing.example.model.RepositoryManager
import com.bing.example.model.entity.VideoInfo
import com.bing.mvvmbase.base.recycleview.BaseRecycleViewFragment
import com.bing.mvvmbase.model.datawrapper.Status
import com.blankj.utilcode.util.FileUtils
import com.blankj.utilcode.util.SizeUtils
import com.blankj.utilcode.util.ToastUtils
import com.scwang.smartrefresh.layout.SmartRefreshLayout
import com.scwang.smartrefresh.layout.api.RefreshLayout
import com.yarolegovich.lovelydialog.LovelyStandardDialog
import com.yarolegovich.lovelydialog.LovelyTextInputDialog

import java.io.File

import com.bing.example.main.videolist.VideosAdapter.Companion.MODE_SELECT

class VideoListFragment : BaseRecycleViewFragment<FragmentVideoListBinding, VideoListViewModel, MainViewModel, VideosAdapter, VideoInfo>() {
        override fun initActivityViewModel() {
                mActivityViewModel = ViewModelProviders.of(activity!!).get(MainViewModel::class.java)
        }

        override fun initViewModel() {
                mViewModel = ViewModelProviders.of(this).get(VideoListViewModel::class.java)
        }


        override val refreshLayout: SmartRefreshLayout
                get() = mBinding.refreshLayout

        override val recyclerView: RecyclerView
                get() = mBinding.recyclerView

        override val layoutManager: RecyclerView.LayoutManager
                get() = GridLayoutManager(context, 2)

        override val networkState: MutableLiveData<Status>
                get() = mViewModel.loadStatus

        override val refreshState: MutableLiveData<Status>
                get() = mViewModel.loadStatus

        override val data: LiveData<List<VideoInfo>>
                get() = mViewModel.videoInfos

        override fun handleArguments() {

        }

        override fun layoutId(): Int {
                return R.layout.fragment_video_list
        }

        override fun initRefreshLayout() {
                super.initRefreshLayout()
                refreshLayout.setEnableRefresh(false)
        }

        override fun refresh(refreshLayout: RefreshLayout) {

        }

        override fun initRecycleView() {
                super.initRecycleView()
                val offsetsItemDecoration = GridOffsetsItemDecoration(
                        GridOffsetsItemDecoration.GRID_OFFSETS_VERTICAL)
                offsetsItemDecoration.setVerticalItemOffsets(SizeUtils.dp2px(8f))
                offsetsItemDecoration.setHorizontalItemOffsets(SizeUtils.dp2px(8f))
                mRecyclerView.addItemDecoration(offsetsItemDecoration)
        }

        override fun initAdapter() {
                mAdapter = VideosAdapter(object : VideosAdapter.OnClickListener {
                        override fun onClickImage(position: Int, videoInfo: VideoInfo) {
                                val fileName = videoInfo.path
                                if (FileUtils.isFileExists(fileName)) {
                                        JzvdStd.startFullscreen(context, JzvdStd::class.java, videoInfo.path, videoInfo.title)
                                } else {
                                        ToastUtils.showShort(R.string.file_not_exits)
                                }
                        }

                        override fun onClickRename(position: Int, videoInfo: VideoInfo) {
                                MaterialDialog(context!!).show {
                                        message(R.string.rename)
                                        input { _, text ->
                                                val videoInfo1 = videoInfo.copy()
                                                videoInfo1.title = text.toString()
                                                RepositoryManager.instance().updateVideo(videoInfo1)
                                        }
                                        positiveButton(android.R.string.ok)
                                }
                        }

                        override fun onClickDelete(position: Int, videoInfo: VideoInfo) {
                                MaterialDialog(context!!).show {
                                        message(R.string.sure_delete)
                                        positiveButton(android.R.string.ok) {
                                                mViewModel.deleteVideo(videoInfo)
                                        }
                                        negativeButton(android.R.string.no) { }
                                }
                        }

                        override fun onClickShare(position: Int, videoInfo: VideoInfo) {
                                val uri = FileProvider.getUriForFile(context!!, "com.bing.example.fileprovider", File(videoInfo.path))

                                val shareIntent = Intent()
                                shareIntent.action = Intent.ACTION_SEND            //分享视频只能单个分享
                                shareIntent.putExtra(Intent.EXTRA_STREAM, uri)
                                shareIntent.type = "audio/*"
                                startActivity(Intent.createChooser(shareIntent, getString(R.string.share_to)))
                        }

                        override fun onClickFore(position: Int, selectedNumber: Int) {

                        }
                }, object : VideosAdapter.OnModeChangeListener {
                        override fun onModeTo(mode: Int) {
                                if (mode == VideosAdapter.MODE_NORMAL) {
                                        mActivityViewModel.isNormalMode.set(true)
                                } else {
                                        mActivityViewModel.isNormalMode.set(false)
                                }
                        }
                })
        }

        fun onClickBack() {
                mAdapter.intoNormalMode()
                mActivityViewModel.isNormalMode.set(true)
        }

        fun onClickDelete() {
                mViewModel.deleteVideos(mAdapter.selectedData)
                mAdapter.intoNormalMode()
                mActivityViewModel.isNormalMode.set(true)
        }

        fun onClickSelectAll() {
                if (mAdapter.selectedPostions.size == mAdapter.itemCount) {
                        mAdapter.selectedPostions.clear()
                } else {
                        for (i in 0 until mAdapter.data!!.size) {
                                mAdapter.selectedPostions.add(i)
                        }
                }
                mAdapter.notifyDataSetChanged()
        }

        fun onBackPressed(): Boolean {
                if (mAdapter.currentMode == MODE_SELECT) {
                        mAdapter.intoNormalMode()
                        return true
                } else {
                        return false
                }
        }

        fun intoNormalMode() {
                if (mAdapter.currentMode == MODE_SELECT) {
                        mAdapter.intoNormalMode()
                }
        }

        override fun initStatusLayout() {
                mStatusLayout = mBinding.statusLayout
        }
}
